File: backends/open-liberty-java/src/main/java/dev/stackverse/openliberty/ReportInput.java
package dev.stackverse.openliberty;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;

/** JSON request DTO for creating and editing reports. */
public record ReportInput(
        @NotBlank(message = "{validation.report.reason.invalid}") @Pattern(
                        regexp = "spam|offensive|broken-link|other",
                        message = "{validation.report.reason.invalid}")
                String reason,
        @Size(max = 1000, message = "{validation.report.comment.too-long}") String comment) {}

File: backends/elixir-phoenix/lib/stackverse_backend/inputs/report_input.ex
defmodule StackverseBackend.Inputs.ReportInput do
  @moduledoc false

  use Ecto.Schema

  import Ecto.Changeset

  alias StackverseBackend.Inputs.Support

  @primary_key false
  embedded_schema do
    field :reason, :string
    field :comment, :string
  end

  @reasons ~w[spam offensive broken-link other]
  @cast_messages %{
    reason: "validation.report.reason.invalid",
    comment: "validation.report.comment.too-long"
  }

  def changeset(body) do
    changeset =
      Support.contract_cast(%__MODULE__{}, body, [:reason, :comment], @cast_messages)

    reason = get_field(changeset, :reason)
    comment = get_field(changeset, :comment)

    changeset
    |> Support.require(
      :reason,
      is_binary(reason) and reason in @reasons,
      "validation.report.reason.invalid"
    )
    |> Support.require(
      :comment,
      Support.length_of(comment) <= 1000,
      "validation.report.comment.too-long"
    )
  end
end
